pages/best-practices/best-practices.mdx

```diff
@@ -0,0 +1,293 @@
+However, this increases the likelihood of drift across your system, hides the authorization logic for a system in that system's codebase, and increases the load on SpiceDB.
+
+### Avoid Cycles in your Schema
+
+Tags: **schema**
+
+Recursive schemas can be very powerful, but can also lead to large performance issues when used incorrectly.
+A good rule of thumb is, if you need a schema definition to recur, have it refer to itself (e.g., groups can have subgroups).
+Avoid situations where a definition points to a separate definition that, further down the permission chain, points to the original definition by accident.
+
+Avoid:
+
+```zed
+definition user {
+	relation org: organization
+}
+
+definition group {
+	relation member: user
+}
+
+definition organization {
+	relation subgroup: group
+}
+```
+
+Preferred:
+
+```zed
+definition user {}
+
+definition group {
+	relation member: user | group
+}
+```
+
+### Phrase Permissions Additively/Positively
+
+Tags: **schema**
+
+A more comprehensible permission system is a more secure permission system.
+One of the easiest ways to maintain your authorization logic is to treat permissions as `positive` or `additive`: a user gains permissions when relations are written.
+This reduces the number of ways that permission logic can interact, and prevents the granting of permission accidentally.
+
+In concrete terms, that means use wildcards and negations sparingly.
+Start with no access and build up; don’t start with full access and pare down.
+
+### Use Unique Identifiers for Object Identifiers
+
+Tags: **application**
+
+Because you typically want to centralize your permissions in SpiceDB, that also means that most of the `IDs` of objects in SpiceDB are references to external entities.
+These external entities shouldn't overlap.
+To that end, we recommend either using `UUIDs` or using another identifier from the upstream that you can be sure will be unique, such as the unique sub field assigned to a user token by your IDP.
+
+### Avoid ReadRelationships API
+
+Tags: **application**
+
+The `ReadRelationships` API should be treated as an escape hatch, used mostly for data introspection.
+Using it for permission logic is a code smell.
+All checks and listing of IDs should use `Check`, `CheckBulk`, `LookupResources`, and `LookupSubjects`.
+If you find yourself reaching for the `ReadRelationships` API for permission logic, there's probably a way to modify your schema to use one of the check APIs instead.
+
+### Prefer CheckBulk To LookupResources
+
+Tags: **application**
+
+Both `CheckBulk` and `LookupResources` can be used to determine whether a subject has access to a list of objects.
+Where possible, we recommend `CheckBulk`, because its work is bounded to the list of requested checks, whereas the wrong `LookupResources` call can return the entire world and choke both your datastore and your client.
```

Comment:
```suggestion
Where possible, we recommend `CheckBulk`, because its work is bounded to the list of requested checks, whereas the wrong `LookupResources` call can return the entire world and therefore be slow.
```